### PR TITLE
Pull annotator out of the sidebar

### PR DESCRIPTION
**Pull request by aron**
Dokku: https://syncbridge.dokku.hypothes.is

This leaves annotator only to serve the Auth plugin. Once that is refactored out, then annotator will be gone.

This is a work in progress commit, that aims to take us towards the discussed architecture in the diagram below.

https://cloud.githubusercontent.com/assets/47144/5860292/bf1c06e4-a260-11e4-81fa-698ea9b673d3.jpg

However we're still a little way from that and I'd like to propose that we get this into a fit state to merge and then work on the remaining changes individually so that they're easier to review.

This PR adds:

* An `annotationUI` service that keeps track of the state shared between the sidebar and the guest frames.
* An `AnnotationUISync` object that replaces `Hypothesis`. This deals with syncing the AnnotationUI state between frames.
* An `AnnotationUIController` that is instantiated in the `AppController` to update the scope accordingly when the annotationUI state changes.
* A `Discovery` object that deals with finding other frames.
* An `AnnotationSync` object that keeps the annotation state sync between the frames and the sidebar.
* A `Bridge` object that acts as a message bus between the frames (was `Bridge`).
* A `crossframe` service that can be used by the application for sending notifications to other frames. It also bootstraps all of the above objects.
* The `Annotator.Plugin.Bridge` has also now been renamed to `Annotator.Plugin.CrossFrame` to keep it in line with the new service.

To be done.

- [x] Pull out the `CrossFrameSync` instantiation from the `AnnotationUISync` service and instantiate them separately as in the diagram. @aron
- [x] Rename CrossFrameCommunication to CrossFrameBridge @Treora 
- [x] Rename CrossFrameSync to AnnotationSync @Treora 
- [x] Update AnnotationUISync to just be a constructor not a service. @aron
- [x] Clean up the `AnnotationUISync` service and add tests. @aron
- [x] Move CrossFrameDiscovery initialization into CrossFrameService @Treora 
- [x] Wire up AnnotationUISync and AnnotationSync in the CrossFrameService @aron
- [x] CamelCase all method names. @Treora 
- [x] Add tests for `AnnotationUIController`. @aron
- [x] Update Annotator.Bridge to use the discovery and communication objects @Treora
- [x] ~~Move the providers/entities code from the controllers into a sync module.~~ _edit: This is bigger than it originally looked and I'm going to suggest we do this as part of a separate pull request._
- [x] Delete `Hypothesis` and `Annotator.Threading`
- [x] Tests for `CrossFrameBridge` @aron
- [x] Tests for `CrossFrameDiscovery`
- [x] Tests for `Annotator.Bridge`
- [x] Tests for `Guest`

Hopefully this makes some sense. Would love thoughts.

### CODE REVIEW

File: h/static/scripts/cross-frame-sync.coffee [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  _parseResults: (results) ->
+    if Array.isArray(results[0])
+      acc = []
+      foldFn = (_, cur) =>
+        (this._parse(a) for a in cur)
+    else
+      acc = {}
+      foldFn = (_, cur) =>
+        this._parse(cur)
+    results.reduce(foldFn, acc)
```
Comment (aron) [inline h/static/scripts/cross-frame-sync.coffee]: @Treora I'm not sure this works as expected, this will result in only the last passed result being returned. I would expect `.map` to be used here.

```coffee
results.map(this._parse, this)
```

Comment (Treora) [inline h/static/scripts/cross-frame-sync.coffee]: I've also been wondering why `reduce` was used. The result is not used so I guess it works anyway, but will change it albeit for readability.

File: h/static/scripts/cross-frame-communication.coffee [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  _create_channel: (source, origin, token) ->
+    # Set up a channel
+    scope = @options.scope + token
+    channelOptions =
+      window: source
+      origin: origin
+      scope: scope
+      onReady: (channel) =>
+        for callback in @onConnectListeners
+          callback.call(this, channel)
```
Comment (aron) [inline h/static/scripts/cross-frame-communication.coffee]: No need for the `this` here, just call it with the channel directly.

**Comment by aron** [comment]
@Treora just rebased on top of master, so beware :) :rotating_light:

**Comment by Treora** [comment]
Updated some things, but I'm not happy yet with the interfaces between the components. Still have to look what would make most sense.

File: h/static/scripts/annotation-mapper-service.coffee [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          this.setupAnnotation(n)
+
+        # If there are more to do, do them after a 10ms break (for browser
+        # responsiveness).
+        if annList.length > 0
+          setTimeout((-> loader(annList)), 10)
+        else
+          $rootScope.$emit('annotationsLoaded', clone)
+
+      clone = annotations.slice()
```
Comment (nickstenning) [inline h/static/scripts/annotation-mapper-service.coffee]: I know this'll be hoisted, but it might be clearer to put it above `loader` for the sake of clarity.

File: h/static/scripts/cross-frame-service.coffee [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+          $rootScope.$emit.call($rootScope, args...)
+        on: (event, handler) ->
+          $rootScope.$on(event, handler)
+
+      new AnnotationSync(options, bridge)
+
+    createAnnotationUISync = (bridge) ->
+      new AnnotationUISync($rootScope, $window, bridge, annotationUI)
+
+    addProvider = (channel) =>
```
Comment (nickstenning) [inline h/static/scripts/cross-frame-service.coffee]: This whole function seems a bit weird. Why is this in here? Why is it not part of one or other of the `*Sync` objects?

File: h/static/scripts/controllers.coffee [hunk 5]
```diff
@@ -170,45 +198,43 @@ class ViewerController
     loaded = []
 
     loadAnnotations = ->
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: Again, this seems like an odd place for this function (and the `$watchCollection` below). It seems to me that this should be part of the responsibilities of `AnnotationSync`.

File: h/static/scripts/controllers.coffee [hunk 6]
```diff
@@ -113,18 +142,18 @@ class AppController
...
 class AnnotationViewerController
   this.$inject = [
     '$location', '$routeParams', '$scope',
-    'annotator', 'streamer', 'store', 'streamfilter'
+    'streamer', 'store', 'streamfilter', 'annotationMapper'
   ]
   constructor: (
      $location,   $routeParams,   $scope,
-     annotator,   streamer,   store,   streamfilter
+     streamer,   store,   streamfilter,   annotationMapper
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: Injecting both `store` and `annotationMapper` seems like an antipattern here. If it's a data mapper then it should abstract away the need to talk to the persistence layer (`store`).

File: h/static/scripts/controllers.coffee [hunk 7]
```diff
@@ -170,45 +198,43 @@ class ViewerController
...
         return unless auth.user
         query = user: auth.user
 
-      for p in annotator.providers
+      for p in crossframe.providers
         for e in p.entities when e not in loaded
           loaded.push e
-          store.SearchResource.get angular.extend(uri: e, query), (results) ->
-            annotator.loadAnnotations(results.rows)
+          r = store.SearchResource.get angular.extend(uri: e, query), (results) ->
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: Ditto vis-à-vis injecting both the `store` and the `annotationMapper`. It seems like this happens because the search resource is bolted onto annotation persistence. Perhaps introduce a distinct `searchMapper` class which wraps the `store` but only exposes a search API.

File: h/static/scripts/controllers.coffee [hunk 8]
```diff
@@ -170,45 +198,43 @@ class ViewerController
...
 
       streamer.send({filter: streamfilter.getFilter()})
 
-    $scope.$watch (-> annotator.tool), (newVal, oldVal) ->
+    $scope.$watch (-> annotationUI.tool), (newVal, oldVal) ->
       return if newVal is oldVal
       $route.reload()
 
-    $scope.$watchCollection (-> annotator.providers), loadAnnotations
+    $scope.$watchCollection (-> crossframe.providers), loadAnnotations
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: As mentioned above, this seems like the wrong place for this. It should be the responsibility of the `AnnotationSync` component to ensure that the right annotations are loaded in all frames.

File: h/static/scripts/controllers.coffee [hunk 9]
```diff
@@ -170,45 +198,43 @@ class ViewerController
...
     $scope.focus = (annotation) ->
       if angular.isObject annotation
         highlights = [annotation.$$tag]
       else
         highlights = []
-      for p in annotator.providers
-        p.channel.notify
-          method: 'focusAnnotations'
-          params: highlights
+      crossframe.notify
```
Comment (nickstenning) [inline h/static/scripts/controllers.coffee]: Yay! I like this simplification.

File: h/static/scripts/cross-frame-bridge.coffee [hunk 10]
```diff
@@ -0,0 +1,109 @@
+class CrossFrameBridge
```
Comment (nickstenning) [inline h/static/scripts/cross-frame-bridge.coffee]: I think I'd prefer it if this component were called just "Bridge". I sure hope we're not going to have multiple types of Bridge, and [longer names aren't always better](http://doc.cat-v.org/bell_labs/pikestyle).

File: h/static/scripts/cross-frame-bridge.coffee [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    @onConnectListeners.push(callback)
+
+
+  # Construct a channel to another frame
+  _buildChannel: (options) ->
+    # jschannel chokes on FF and Chrome extension origins.
+    if (options.origin.match /^chrome-extension:\/\//) or
+        (options.origin.match /^resource:\/\//)
+      options = $.extend {}, options, {origin: '*'}
+    channel = Channel.build(options)
```
Comment (nickstenning) [inline h/static/scripts/cross-frame-bridge.coffee]: Are we relying on our current concatenate-before-compile workflow to expose this component to others? That seems... risky. This seems like a reasonable candidate for an angular service.

File: h/static/scripts/cross-frame-discovery.coffee [hunk 12]
```diff
@@ -0,0 +1,108 @@
+class CrossFrameDiscovery
```
Comment (nickstenning) [inline h/static/scripts/cross-frame-discovery.coffee]: Again, I think I prefer the terser `Discovery`, or possibly `FrameDiscovery` at a push.

File: h/static/scripts/cross-frame-discovery.coffee [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        # Request it to set up a channel if we did not already do so.
+        unless @requestInProgress?
+          @requestInProgress = true # prevent creating two channels
+          reply = 'request'
+      else if messageType is 'ack'
+        # The other side opened a channel to us. We note its scope and create
+        # a matching channel end on this side.
+        @requestInProgress = false # value should not actually matter anymore.
+        discovered = true
+    return {reply: reply, discovered: discovered, token: token}
```
Comment (nickstenning) [inline h/static/scripts/cross-frame-discovery.coffee]: Ditto on the subject of how this component is exposed to the rest of the code. An angular service here?

File: h/static/scripts/cross-frame-bridge.coffee (hunk 11 above)
Comment (aron) [inline h/static/scripts/cross-frame-bridge.coffee]: I've started exporting them as values for testing. Only the CrossFrameService should be a service for use across the application.

File: h/static/scripts/directives/annotation.coffee [hunk 14]
```diff
@@ -26,14 +26,15 @@ validate = (value) ->
 #
 # `AnnotationController` provides an API for the annotation directive. It
 # manages the interaction between the domain and view models and uses the
-# {@link annotator annotator service} for persistence.
+# {@link annotationMapper service} for persistence.
```
Comment (nickstenning) [inline h/static/scripts/directives/annotation.coffee]: The format of ngDoc @link is `{@link thingToLinkTo name to link to}` so I think you mean `{@link annotationMapper AnnotationMapper service}` or similar here.

File: h/static/scripts/streamsearch.coffee [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
-    'streamer', 'streamfilter'
+    'auth', 'queryparser', 'searchfilter', 'store',
+    'streamer', 'streamfilter', 'annotationMapper'
   ]
   constructor: (
      $scope,   $rootScope,   $routeParams
-     annotator,   auth,   queryparser,   searchfilter,   store,
-     streamer,   streamfilter
+     auth,   queryparser,   searchfilter,   store,
+     streamer,   streamfilter, annotationMapper
```
Comment (nickstenning) [inline h/static/scripts/streamsearch.coffee]: You've replaced `annotator` with `annotationMapper`, but haven't updated the code below. Also, see earlier comments about injecting both `store` and `annotationMapper`.

File: h/static/scripts/threading-service.coffee [hunk 16]
```diff
@@ -0,0 +1,97 @@
+class ThreadingService
+  # Mix in message thread properties into the prototype. The body of the
+  # class will overwrite any methods applied here. If you need inheritance
+  # assign the message thread to a local varible.
+  $.extend(this.prototype, mail.messageThread())
```
Comment (nickstenning) [inline h/static/scripts/threading-service.coffee]: It's not immediately obvious to me where `mail` came from. Could you add a comment pointing the reader to `jwz.js`? I know this will be clearer once we browserify things, but for the moment it's a bit magic.

File: karma.config.js [hunk 17]
```diff
@@ -36,7 +38,6 @@ module.exports = function(config) {
       'h/static/scripts/vendor/annotator.js',
       'h/static/scripts/annotator/monkey.js',
       'h/static/scripts/vendor/annotator.auth.js',
-      'h/static/scripts/vendor/polyfills/url.js',
       'h/static/scripts/annotator/plugin/bridge.js',
       'h/static/scripts/annotator/plugin/bucket-bar.js',
       'h/static/scripts/annotator/plugin/threading.js',
```
Comment (nickstenning) [inline karma.config.js]: This can presumably now go?

File: h/static/scripts/controllers.coffee (hunk 7 above)
Comment (aron) [inline h/static/scripts/controllers.coffee]: I agree, although this PR is already huge and I'd like to avoid adding any more unrelated changes than strictly necessary.

File: h/static/scripts/cross-frame-discovery.coffee (hunk 13 above)
Comment (aron) [inline h/static/scripts/cross-frame-discovery.coffee]: Yup, although it'll be shared between the guest and the sidebar so we'll have to sniff for the angular object.

File: h/static/scripts/annotation-sync.coffee [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+          this._parse(a) unless a is null
+      else
+        # 'annotations' is in fact just a single annotation
+        this._parse(annotations) unless annotations is null
+    return
+
+  # Assign a non-enumerable tag to objects which cross the bridge.
+  # This tag is used to identify the objects between message.
+  _tag: (msg, tag) ->
+    return msg if msg.$$tag
```
Comment (Treora) [inline h/static/scripts/annotation-sync.coffee]: When exactly does this happen? And is it certain that `@cache` does not need updating in that situation?

**Comment by aron** [comment]
@Treora hey, this is getting pretty close to working now. The current issue I've hit is that the discovery system is discovering each frame twice, once in app.js and once in bridge.js. This creates two channels between the two frames which causes messages, like "beforeAnnotationCreated" to be sent twice.

I'm currently trying to work out where the logic is/was to de-dupe frames but if you get a chance to take a look I'd appreciate the help.

**Comment by Treora** [comment]
> The current issue I've hit is that the discovery system is discovering each frame twice

The problem appears to be that `CrossFrameService.connect()` is called twice, which then sets up the whole cross-frame stuff twice. I don't know enough about how Angular runs services to reason about this.

**Comment by aron** [comment]
@Treora thank you!

**Comment by aron** [comment]
It was the assets.yaml including the file twice.

**Comment by aron** [comment]
OH MY! THE TESTS ARE PASSING! :green_heart:

File: h/static/scripts/annotation-sync.coffee [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+    annotations = (this._format a for a in annotations)
+    @bridge.call
+      method: 'sync'
+      params: annotations
+      callback: cb
+    this
+
+  # Handlers for messages arriving through a channel
+  _channelListeners:
+    'beforeCreateAnnotation': (txn, body) ->
```
Comment (nickstenning) [inline h/static/scripts/annotation-sync.coffee]: Lots of repetition here. Could this be replaced with:

```coffeescript
'beforeCreateAnnotation': annotationEventProxy(['beforeAnnotationCreated'])
'createAnnotation': annotationEventProxy(['annotationCreated'])
'updateAnnotation': annotationEventProxy(['beforeAnnotationUpdated', 'annotationUpdated'])
'deleteAnnotation': annotationEventProxy(['annotationDeleted'])

...

annotationEventProxy = (events) ->
  (txn, body) ->
    annotation = this._parse(body)
    delete @cache[annotation.$$tag]
    for e in events
      @_emit(e, annotation)
    @cache[annotation.$$tag] = annotation
    this._format annotation
```

Comment (Treora) [inline h/static/scripts/annotation-sync.coffee]: Indeed. There's this and a few more things that I would like to refactor, but first let's get everything working.

**Comment by aron** [comment]
Getting there slowly, @Treora if you get a moment next week could you look into adding tests for `CrossFrameDiscovery`, I'm starting the `CrossFrameBridge` now.

**Comment by tilgovi** [comment]
:heart_eyes:

File: h/static/scripts/cross-frame-discovery.coffee [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        token = ':' + ('' + Math.random()).replace(/\D/g, '')
+        reply = 'ack' + token
+        discovered = true
+    else # We are configured as a client
+      if messageType is 'offer'
+        # The server joined the party, or replied to our discovery message.
+        # Request it to set up a channel if we did not already do so.
+        unless @requestInProgress?
+          @requestInProgress = true # prevent creating two channels
+          reply = 'request'
```
Comment (aron) [inline h/static/scripts/cross-frame-discovery.coffee]: @Treora Just to clarify the intent here... a "client" can only establish a connection to a single "server"? Otherwise  `@requestInProgress` will need to be more than a boolean flag as this assumes it's only receiving messages from a single source.

Comment (tilgovi) [inline h/static/scripts/cross-frame-discovery.coffee]: Yes. @Treora and I discussed this in Budapest.

Comment (aron) [inline h/static/scripts/cross-frame-discovery.coffee]: @tilgovi thanks, I'll add a note to this effect.

File: h/static/scripts/cross-frame-discovery.coffee [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+        queue.push child
+    return
+
+
+  _onMessage: (e) =>
+    {source, origin, data} = e.originalEvent
+
+    # Fix for local testing (needed at least in Firefox 34.0)
+    if origin is 'null'
+      origin = '*'
```
Comment (aron) [inline h/static/scripts/cross-frame-discovery.coffee]: I can't find any history for this? Could someone explain what it's for, and what local testing means.

Comment (tilgovi) [inline h/static/scripts/cross-frame-discovery.coffee]: This does not appear in the latest bridge code on master. I'm not familiar with this at all. My guess is that the origin is null if you're testing on a file: URI or something. Kill it.

Comment (aron) [inline h/static/scripts/cross-frame-discovery.coffee]: Good good, @Treora feel free to add this back with some more detailed docs, if it is actually needed in production/development.

Comment (Treora) [inline h/static/scripts/cross-frame-discovery.coffee]: I indeed added it as a quick fix to test on local files. Fine with me to kill it again, I don't even know if the issue is systematic or browser-dependent.

**Comment by aron** [comment]
@Treora just FYI, I'm starting on the `Annotator.Guest` tests.

**Comment by aron** [comment]
@Treora just starting the tests for `Bridge`

**Comment by aron** [comment]
Okay, renamed everything in accordance with @nickstenning's suggestions, we now only use CrossFrame for the service (angular application interface) and for the annotator plugin (guest page interface)

**Comment by aron** [comment]
Been testing this all afternoon, ran into a few issues with rendering updates, fixed those and now seems fine. Could a few other people check the dokku and +1 this PR.

**Comment by gergely-ujvari** [comment]
OMG, this is the biggest PR we every had.

**Comment by gergely-ujvari** [comment]
I was trying to register a new user on the dokku and got this:
```
 Failed to load resource: the server responded with a status of 500 (Internal Server Error)
angular.min.js:4 SyntaxError: Unexpected token I
    at Object.parse (native)
    at Object.fromJson (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:1:10348)
    at process (https://syncbridge.dokku.hypothes.is/assets/scripts/session.min.js?36bd71c5:1:676)
    at transformData (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:3:18705)
    at transformResponse (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:3:22045)
    at wrappedErrback (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:5:735)
    at https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:5:2320
    at Scope.$eval (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:5:11082)
    at Scope.$digest (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:5:8886)
    at Scope.$apply (https://syncbridge.dokku.hypothes.is/assets/scripts/vendor/angular.min.js?afd85d02:5:11515)
```

**Comment by tilgovi** [comment]
That is probably not new. Hard to say the root cause without looking at
server logs but I think we have a know, open issue for the session service
not handling unexpected or error responses gracefully (in this case,
unexpectedly not JSON responses).

**Comment by aron** [comment]
@gergely-ujvari this is the user table creating issue. I'll try and fix it, for the moment could you just pull the branch down and test locally? Thanks.

**Comment by aron** [comment]
Right, it's fixed again now.

**Comment by gergely-ujvari** [comment]
@aron: Right now it's working.

**Comment by gergely-ujvari** [comment]
I'm seeing an issue I'm not seeing on my local master branch, nor on my prod extension.

Steps:
1. Use the syncbridge bookmarklet on two different tabs within the same browsers (i.e I used it on https://en.wikipedia.org/wiki/Global_warming and http://hypothes.is/ )
2. Create an annotation on one of these pages. (I created https://syncbridge.dokku.hypothes.is/a/AUtZxiG5SbjD6_Qdwr8z)
3. Check the other page.

Expected result: The created annotation doesn't appear in this tab.
Actual result: It appears.

global_warming

**Comment by gergely-ujvari** [comment]
Otherwise it is working fine.

**Comment by aron** [comment]
@gergely-ujvari thanks, I'll have a look

**Comment by tilgovi** [comment]
Checked it out. Looks great. Couldn't reproduce any problems described in this thread. Rebasing was tricky, so I merged master in instead. I made some small tweaks to make digests simpler and make the AnnotationUIController a bit less special.

**Comment by tilgovi** [comment]
Merged. All of the :hearts: and all of the :fireworks:. Thanks for all the work on this, @aron and @Treora.

**Comment by Treora** [comment]
Awesome. My pleasure!

**Comment by aron** [comment]
:fireworks:
